File: mokksy/src/jvmTest/java/dev/mokksy/mokksy/MokksyServerJavaLifecycleIT.java
package dev.mokksy.mokksy;

import dev.mokksy.Mokksy;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.NoSuchElementException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class MokksyServerJavaLifecycleIT {
    @Test
    public void startAndShutdownWorkAsBlockingLifecycleMethods() {
        var mokksy = Mokksy.create().start();

        try {
            assertThat(mokksy.port()).isGreaterThan(0);
            assertThat(mokksy.baseUrl()).contains("http://");
            assertThat(mokksy.baseUrl()).contains(String.valueOf(mokksy.port()));
        } finally {
            mokksy.shutdown();
        }
    }

    @Test
    public void closeShutsDownTheServer() {
        try (var mokksy = Mokksy.create().start()) {
            assertThat(mokksy.port()).isGreaterThan(0);
        }
    }

    @Test
    public void shutdownWithCustomTimingsCompletesWithoutError() {
        var mokksy = Mokksy.create().start();
        try {
            mokksy.shutdown(100L, 200L);
        } finally {
            mokksy.shutdown();
        }
    }

    @Test
    public void serverRespondsAfterStart() throws IOException, InterruptedException {
        var mokksy = Mokksy.create().start();

        var httpClient = HttpClient.newHttpClient();
        try {
            mokksy.get(spec -> spec.path("/java-lifecycle-test"))
                .respondsWith(rb -> rb.body("alive"));

            var response = httpClient.send(
                HttpRequest.newBuilder()
                    .uri(URI.create(mokksy.baseUrl() + "/java-lifecycle-test"))
                    .GET()
                    .build(),
                HttpResponse.BodyHandlers.ofString()
            );

            assertThat(response.statusCode()).isEqualTo(200);
            assertThat(response.body()).isEqualTo("alive");
        } finally {
            mokksy.shutdown();
        }
    }

    @Test
    public void getStubReturnsNamedStubHandle() {
        var mokksy = Mokksy.create().start();

        try {
            var expected = mokksy.get(new StubConfiguration("java-get-stub"), "/java-get-stub")
                .respondsWith("alive");

            assertThat(mokksy.getStub("java-get-stub").getName()).isEqualTo(expected.getName());
        } finally {
            mokksy.shutdown();
        }
    }

    @Test
    public void getStubThrowsWhenStubNotFound() {
        var mokksy = Mokksy.create().start();

        try {
            assertThatThrownBy(() -> mokksy.getStub("missing-stub"))
                .isInstanceOf(NoSuchElementException.class)
                .hasMessage("No stub registered with name 'missing-stub'");
        } finally {
            mokksy.shutdown();
        }
    }
}

File: mokksy/src/jvmTest/kotlin/dev/mokksy/mokksy/MokksyServerJavaLifecycleIT.kt
package dev.mokksy.mokksy

import dev.mokksy.Mokksy
import io.kotest.assertions.assertSoftly
import io.kotest.matchers.ints.shouldBeGreaterThan
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.ktor.client.request.get
import io.ktor.client.statement.bodyAsText
import io.ktor.http.HttpStatusCode
import org.junit.jupiter.api.Test

internal class MokksyServerJavaLifecycleIT {
    @Test
    fun `start and shutdown work as blocking lifecycle methods`() {
        val mokksy = Mokksy.create().start()

        try {
            assertSoftly {
                mokksy.port() shouldBeGreaterThan 0
                mokksy.baseUrl() shouldContain "http://"
                mokksy.baseUrl() shouldContain "${mokksy.port()}"
            }
        } finally {
            mokksy.shutdown()
        }
    }

    @Test
    fun `close shuts down the server`() {
        val mokksy = Mokksy.create().start()
        mokksy.port() shouldBeGreaterThan 0
        mokksy.close() // must not throw
    }

    @Test
    fun `shutdown with custom timings completes without error`() {
        val mokksy = Mokksy.create().start()

        mokksy.shutdown(gracePeriodMillis = 100, timeoutMillis = 200)
    }

    @Test
    fun `MokksyServer start extension blocks until port is bound and returns server`() {
        val server = MokksyServer()
        val returned = server.start()
        try {
            assertSoftly {
                returned shouldBe server
                server.port() shouldBeGreaterThan 0
            }
        } finally {
            server.shutdown()
        }
    }

    @Test
    suspend fun `server responds after start via createKtorClient`() {
        val mokksy = Mokksy.create().start()
        val client = createKtorClient(mokksy.port())

        try {
            mokksy.get {
                path("/java-lifecycle-test")
            } respondsWith {
                body = "alive"
            }

            val result = client.get("/java-lifecycle-test")
            val responseBody = result.bodyAsText()

            assertSoftly {
                result.status shouldBe HttpStatusCode.OK
                responseBody shouldBe "alive"
            }
        } finally {
            client.close()
            mokksy.shutdown()
        }
    }
}
